File: doodlestudio/numbers.py
"""Display text -> spoken text: numbers become words, nothing else changes.

Only number tokens are rewritten, so clause punctuation (which drives caption
cues) is identical in both strings. Every rewrite is recorded as a span so a
position in the display text can be mapped to the spoken text.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from decimal import Decimal

import cn2an
from num2words import num2words

NUM = r'\d{1,3}(?:,\d{3})+(?:\.\d+)?|\d+(?:\.\d+)?'


@dataclass
class Normalized:
    display: str
    spoken: str
    spans: list = field(default_factory=list)      # (display_start, display_end, spoken_start, spoken_end)

    def to_spoken(self, pos: int) -> int:
        """Spoken-text offset of display offset ``pos`` (start of a rewritten token maps to its start)."""
        shift = 0
        for d0, d1, s0, s1 in self.spans:
            if pos < d0:
                break
            if pos < d1:
                return s0
            shift = s1 - d1
        return pos + shift

    def find(self, phrase: str) -> str | None:
        """The spoken form of a display substring (None when it is not in the display text)."""
        i = self.display.find(phrase)
        if i < 0:
            return None
        return self.spoken[self.to_spoken(i):self.to_spoken(i + len(phrase)) if i + len(phrase) < len(self.display)
                           else len(self.spoken)].strip()


def _latin(ch: str) -> bool:
    return ch.isascii() and ch.isalnum()


def _rewrite(text: str, pattern: re.Pattern, speak) -> Normalized:
    out, spans, last = [], [], 0
    for m in pattern.finditer(text):
        words = speak(m)
        if words is None:
            continue
        if _latin(text[m.start() - 1:m.start()]) and _latin(words[:1]):     # keep words apart: B2B -> B two B
            words = ' ' + words
        if _latin(text[m.end():m.end() + 1]) and _latin(words[-1:]):
            words += ' '
        out.append(text[last:m.start()])
        s0 = sum(map(len, out))
        out.append(words)
        spans.append((m.start(), m.end(), s0, s0 + len(words)))
        last = m.end()
    out.append(text[last:])
    return Normalized(text, ''.join(out), spans)


# ============================================================== English
MONTHS = {m[:3].lower(): m for m in ('January', 'February', 'March', 'April', 'May', 'June', 'July', 'August',
                                     'September', 'October', 'November', 'December')}
SCALES = {'k': 'thousand', 'thousand': 'thousand', 'm': 'million', 'mn': 'million', 'mm': 'million',
          'million': 'million', 'b': 'billion', 'bn': 'billion', 'billion': 'billion', 't': 'trillion',
          'tn': 'trillion', 'trillion': 'trillion'}
CURRENCIES = {'US$': ('dollar', 'dollars'), '$': ('dollar', 'dollars'), '€': ('euro', 'euros'),
              '£': ('pound', 'pounds'), '¥': ('yen', 'yen'), '₹': ('rupee', 'rupees')}
UNITS = {'km/h': 'kilometers per hour', 'mph': 'miles per hour', 'km': 'kilometers', 'kg': 'kilograms',
         'cm': 'centimeters', 'mm': 'millimeters', '°C': 'degrees Celsius', '°F': 'degrees Fahrenheit',
         '°': 'degrees', 'GB': 'gigabytes', 'MB': 'megabytes', 'TB': 'terabytes', 'GHz': 'gigahertz',
         'kWh': 'kilowatt hours', 'MW': 'megawatts', 'GW': 'gigawatts', 'lbs': 'pounds', 'ft': 'feet'}
FRACTIONS = {'1/2': 'one half', '1/3': 'one third', '2/3': 'two thirds', '1/4': 'one quarter',
             '3/4': 'three quarters', '1/5': 'one fifth', '1/10': 'one tenth'}
SINGULAR = {'kilometers': 'kilometer', 'kilograms': 'kilogram', 'centimeters': 'centimeter',
            'millimeters': 'millimeter', 'degrees': 'degree', 'gigabytes': 'gigabyte', 'megabytes': 'megabyte',
            'terabytes': 'terabyte', 'megawatts': 'megawatt', 'gigawatts': 'gigawatt', 'pounds': 'pound', 'feet': 'foot'}

_cur = '|'.join(re.escape(c) for c in CURRENCIES)
_scale = r'trillion|billion|million|thousand|tn|bn|mn|mm|[kmbt]'
_month = (r'\b(?:January|February|March|April|May|June|July|August|September|October|November|December'
          r'|Jan|Feb|Mar|Apr|Jun|Jul|Aug|Sept|Sep|Oct|Nov|Dec)\.?')
_unit = '|'.join(re.escape(u) for u in sorted(UNITS, key=len, reverse=True))
EN_PATTERN = re.compile(
    rf'(?P<cur>{_cur})\s?(?P<camt>{NUM})(?:\s?(?P<cscale>{_scale})\b)?'
    rf'|(?P<ra>{NUM})\s?(?:-|–|to)\s?(?P<rb>{NUM})\s?(?P<rpct>%)'
    rf'|(?P<pct>-?(?:{NUM}))\s?%'
    rf'|(?P<month>{_month})\s(?P<day>\d{{1,2}})(?!\d|,\d)(?:st|nd|rd|th)?'
    rf'|(?P<h>\d{{1,2}}):(?P<mi>\d{{2}})(?:\s?(?P<ampm>[ap])\.?m\.?)?'
    rf'|(?P<ord>\d+)(?:st|nd|rd|th)\b'
    rf'|(?P<decade>1[1-9]\d0|20[0-9]0)s\b'
    rf'|(?P<ya>1[1-9]\d{{2}}|20\d{{2}})\s?(?:-|–)\s?(?P<yb>1[1-9]\d{{2}}|20\d{{2}})(?!\d)'
    rf'|(?P<mult>{NUM})\s?[x×](?![a-z])'
    rf'|(?<![A-Za-z])(?P<samt>{NUM})(?P<sscale>bn|mn|tn|[kmbKMB])\b'
    rf'|(?P<uamt>{NUM})\s?(?P<unit>{_unit})(?![A-Za-z])'
    rf'|#(?P<hash>\d+)'
    rf'|(?P<frac>\d+/\d+)'
    rf'|(?P<ra2>{NUM})\s?–\s?(?P<rb2>{NUM})'
    rf'|(?P<year>(?<![\d.,$])(?:1[1-9]\d{{2}}|20\d{{2}})(?![\d%]|\.\d|,\d))'
    rf'|(?P<neg>(?<![\w.])-)?(?P<num>{NUM})'
)


def _plain(text: str) -> str:
    return re.sub(r',| and(?= )', '', text)


def en_number(token: str) -> str:
    value = Decimal(token.replace(',', ''))
    return _plain(num2words(value if value % 1 else int(value)))


def en_year(token: str) -> str:
    return _plain(num2words(int(token), to='year'))


def _en_speak(m: re.Match) -> str:
    g = m.groupdict()
    if g['cur']:
        one, many = CURRENCIES[g['cur']]
        amount = g['camt']
        if g['cscale']:
            return f"{en_number(amount)} {SCALES[g['cscale'].lower()]} {many}"
        if re.fullmatch(r'[\d,]+\.\d\d', amount) and not amount.endswith('.00') and one == 'dollar':
            whole, cents = amount.replace(',', '').split('.')
            unit = one if int(whole) == 1 else many
            return f"{en_number(whole)} {unit} and {en_number(cents)} cents"
        return f"{en_number(amount)} {one if Decimal(amount.replace(',', '')) == 1 else many}"
    if g['rpct']:
        return f"{en_number(g['ra'])} to {en_number(g['rb'])} percent"
    if g['pct']:
        token = g['pct']
        return ('minus ' if token.startswith('-') else '') + f"{en_number(token.lstrip('-'))} percent"
    if g['month']:
        month = MONTHS[g['month'].rstrip('.').lower()[:3]]
        return f"{month} {num2words(int(g['day']), to='ordinal')}"
    if g['h']:
        hour, minute = int(g['h']), int(g['mi'])
        if hour > 24 or minute > 59:
            return None
        words = en_number(str(hour)) + (" o'clock" if minute == 0 and not g['ampm'] else
                                          '' if minute == 0 else
                                          f" oh {en_number(str(minute))}" if minute < 10 else f" {en_number(str(minute))}")
        return words + (f" {g['ampm']} m" if g['ampm'] else '')
    if g['ord']:
        return num2words(int(g['ord']), to='ordinal')
    if g['decade']:
        words = en_year(g['decade'])
        return words[:-1] + 'ies' if words.endswith('y') else words + 's'
    if g['ya']:
        return f"{en_year(g['ya'])} to {en_year(g['yb'])}"
    if g['mult']:
        return f"{en_number(g['mult'])} times"
    if g['samt']:
        return f"{en_number(g['samt'])} {SCALES[g['sscale'].lower()]}"
    if g['uamt']:
        first, _, rest = UNITS[g['unit']].partition(' ')
        if Decimal(g['uamt'].replace(',', '')) == 1:
            first = SINGULAR.get(first, first)
        return f"{en_number(g['uamt'])} {first}{' ' + rest if rest else ''}"
    if g['hash']:
        return f"number {en_number(g['hash'])}"
    if g['frac']:
        a, b = g['frac'].split('/')
        return FRACTIONS.get(g['frac'], f'{en_number(a)} {en_number(b)}')
    if g['ra2']:
        return f"{en_number(g['ra2'])} to {en_number(g['rb2'])}"
    if g['year']:
        return en_year(g['year'])
    return ('minus ' if g['neg'] else '') + en_number(g['num'])


def normalize_en(display: str) -> Normalized:
    return _rewrite(display, EN_PATTERN, _en_speak)


# ============================================================== Chinese
ZH_CUR = {'US$': '美元', '$': '美元', '€': '欧元', '£': '英镑', '¥': '元', '￥': '元', 'HK$': '港元'}
_zcur = '|'.join(re.escape(c) for c in sorted(ZH_CUR, key=len, reverse=True))
ZH_PATTERN = re.compile(
    rf'(?P<h>\d{{1,2}})[:：](?P<mi>\d{{2}})'
    rf'|(?P<ra>{NUM})\s?(?:-|–|~|～|至|到)\s?(?P<rb>{NUM})\s?(?P<rpct>[%％])'
    rf'|(?P<cur>{_zcur})\s?(?P<camt>{NUM})(?:\s?(?P<cscale>万亿|亿|万|千|trillion|billion|million|bn|mn|[kmb]))?'
    rf'|(?P<year>\d{{4}})(?=\s?年)'
    rf'|(?P<pct>{NUM})\s?[%％]'
    rf'|(?P<frac>\d+)/(?P<den>\d+)'
    rf'|(?P<num>{NUM})'
)
ZH_SCALES = {'trillion': '万亿', 'billion': '十亿', 'million': '百万', 'bn': '十亿', 'mn': '百万', 'k': '千',
             'm': '百万', 'b': '十亿'}


def zh_number(token: str) -> str:
    words = cn2an.an2cn(token.replace(',', ''), 'low')
    return re.sub(r'(^|[亿万])二(?=[千万亿])', r'\1两', words)


def _zh_speak(m: re.Match) -> str:
    g = m.groupdict()
    if g['h']:
        hour, minute = int(g['h']), int(g['mi'])
        if hour > 24 or minute > 59:
            return None
        return zh_number(str(hour)) + '点' + ('' if minute == 0 else '半' if minute == 30 else
                                              ('零' if minute < 10 else '') + zh_number(str(minute)) + '分')
    if g['rpct']:
        return f"百分之{zh_number(g['ra'])}到百分之{zh_number(g['rb'])}"
    if g['cur']:
        scale = g['cscale'] or ''
        scale = ZH_SCALES.get(scale.lower(), scale)
        return f"{zh_number(g['camt'])}{scale}{ZH_CUR[g['cur']]}"
    if g['year']:
        return ''.join('零一二三四五六七八九'[int(d)] for d in g['year'])
    if g['pct']:
        return f"百分之{zh_number(g['pct'])}"
    if g['frac']:
        return f"{zh_number(g['den'])}分之{zh_number(g['frac'])}"
    return zh_number(g['num'])


def normalize_zh(display: str) -> Normalized:
    return _rewrite(display, ZH_PATTERN, _zh_speak)


def normalize(display: str, lang: str) -> Normalized:
    return normalize_en(display) if lang == 'en' else normalize_zh(display)

File: edu/extension/engine/numbers.js
// Display text -> spoken text: numbers become words, nothing else changes (port of doodlestudio/numbers.py, English).
//
// Only number tokens are rewritten, so clause punctuation (which drives caption cues) is identical in both
// strings. Every rewrite is recorded as a span so a position in the display text maps to the spoken text.

const NUM = String.raw`\d{1,3}(?:,\d{3})+(?:\.\d+)?|\d+(?:\.\d+)?`;

export class Normalized {
  constructor(display, spoken, spans = []) {
    this.display = display;
    this.spoken = spoken;
    this.spans = spans;              // [displayStart, displayEnd, spokenStart, spokenEnd]
  }

  /** Spoken-text offset of display offset `pos` (the start of a rewritten token maps to its start). */
  toSpoken(pos) {
    let shift = 0;
    for (const [d0, d1, s0, s1] of this.spans) {
      if (pos < d0) break;
      if (pos < d1) return s0;
      shift = s1 - d1;
    }
    return pos + shift;
  }

  /** The spoken form of a display substring (null when it is not in the display text). */
  find(phrase) {
    const i = this.display.indexOf(phrase);
    if (i < 0) return null;
    const end = i + phrase.length < this.display.length ? this.toSpoken(i + phrase.length) : this.spoken.length;
    return this.spoken.slice(this.toSpoken(i), end).trim();
  }
}

const latin = (ch) => /^[A-Za-z0-9]$/.test(ch);

function rewrite(text, pattern, speak) {
  const out = [];
  const spans = [];
  let last = 0;
  let length = 0;
  for (const m of text.matchAll(pattern)) {
    let words = speak(m);
    if (words == null) continue;
    const start = m.index;
    const end = start + m[0].length;
    if (latin(text.slice(start - 1, start)) && latin(words.slice(0, 1))) words = ' ' + words;   // B2B -> B two B
    if (latin(text.slice(end, end + 1)) && latin(words.slice(-1))) words += ' ';
    out.push(text.slice(last, start));
    length += start - last;
    out.push(words);
    spans.push([start, end, length, length + words.length]);
    length += words.length;
    last = end;
  }
  out.push(text.slice(last));
  return new Normalized(text, out.join(''), spans);
}

// ------------------------------------------------------------ number words (num2words, English)
const SMALL = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', 'eleven',
  'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'];
const TENS = ['', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'];
const SCALES = [[10n ** 18n, 'quintillion'], [10n ** 15n, 'quadrillion'], [10n ** 12n, 'trillion'],
  [10n ** 9n, 'billion'], [10n ** 6n, 'million'], [1000n, 'thousand']];
const ORDS = { one: 'first', two: 'second', three: 'third', four: 'fourth', five: 'fifth', six: 'sixth',
  seven: 'seventh', eight: 'eighth', nine: 'ninth', ten: 'tenth', eleven: 'eleventh', twelve: 'twelfth' };

/** num2words' cardinal, with its "and" and commas ("one thousand, two hundred and thirty-four"). */
function cardinalFull(n) {
  n = BigInt(n);
  if (n < 20n) return SMALL[Number(n)];
  if (n < 100n) {
    const r = Number(n % 10n);
    return r ? `${TENS[Number(n / 10n)]}-${SMALL[r]}` : TENS[Number(n / 10n)];
  }
  if (n < 1000n) {
    const head = `${SMALL[Number(n / 100n)]} hundred`;
    const r = n % 100n;
    return r ? `${head} and ${cardinalFull(r)}` : head;
  }
  for (const [s, name] of SCALES) {
    if (n >= s) {
      const head = `${cardinalFull(n / s)} ${name}`;
      const r = n % s;
      if (!r) return head;
      return r < 100n ? `${head} and ${cardinalFull(r)}` : `${head}, ${cardinalFull(r)}`;
    }
  }
  throw new Error(`number too large: ${n}`);
}

const plain = (text) => text.replace(/,| and(?= )/g, '');

function ordinal(n) {
  const words = cardinalFull(n).split(' ');
  const parts = words[words.length - 1].split('-');
  let lastWord = parts[parts.length - 1];
  if (ORDS[lastWord]) lastWord = ORDS[lastWord];
  else lastWord = (lastWord.endsWith('y') ? lastWord.slice(0, -1) + 'ie' : lastWord) + 'th';
  parts[parts.length - 1] = lastWord;
  words[words.length - 1] = parts.join('-');
  return words.join(' ');
}

/** num2words for a decimal: "three point one four" (the digits of the float as Python prints it). */
function decimalWords(token) {
  const value = Number(token);
  const text = String(value);                  // like Python's str(float): '3.80' -> '3.8'
  const [whole, frac = ''] = text.split('.');
  return [cardinalFull(BigInt(whole)), ...(frac ? ['point', ...[...frac].map((d) => SMALL[+d])] : [])].join(' ');
}

export function enNumber(token) {
  const clean = token.replace(/,/g, '');
  const [whole, frac = ''] = clean.split('.');
  if (/[1-9]/.test(frac)) return plain(decimalWords(clean));
  return plain(cardinalFull(BigInt(whole)));
}

export function enYear(token) {
  const val = Number(token);
  const high = Math.floor(val / 100);
  const low = val % 100;
  if (high === 0 || (high % 10 === 0 && low < 10) || high >= 100) return plain(cardinalFull(val));
  const lowText = low === 0 ? 'hundred' : low < 10 ? `oh-${cardinalFull(low)}` : cardinalFull(low);
  return plain(`${cardinalFull(high)} ${lowText}`);
}

// ---------------------------------------------------------------- English patterns
const MONTHS = Object.fromEntries(['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August',
  'September', 'October', 'November', 'December'].map((m) => [m.slice(0, 3).toLowerCase(), m]));
const SCALE_WORDS = { k: 'thousand', thousand: 'thousand', m: 'million', mn: 'million', mm: 'million',
  million: 'million', b: 'billion', bn: 'billion', billion: 'billion', t: 'trillion', tn: 'trillion',
  trillion: 'trillion' };
const CURRENCIES = { 'US$': ['dollar', 'dollars'], '$': ['dollar', 'dollars'], '€': ['euro', 'euros'],
  '£': ['pound', 'pounds'], '¥': ['yen', 'yen'], '₹': ['rupee', 'rupees'] };
const UNITS = { 'km/h': 'kilometers per hour', mph: 'miles per hour', km: 'kilometers', kg: 'kilograms',
  cm: 'centimeters', mm: 'millimeters', '°C': 'degrees Celsius', '°F': 'degrees Fahrenheit', '°': 'degrees',
  GB: 'gigabytes', MB: 'megabytes', TB: 'terabytes', GHz: 'gigahertz', kWh: 'kilowatt hours', MW: 'megawatts',
  GW: 'gigawatts', lbs: 'pounds', ft: 'feet' };
const FRACTIONS = { '1/2': 'one half', '1/3': 'one third', '2/3': 'two thirds', '1/4': 'one quarter',
  '3/4': 'three quarters', '1/5': 'one fifth', '1/10': 'one tenth' };
const SINGULAR = { kilometers: 'kilometer', kilograms: 'kilogram', centimeters: 'centimeter',
  millimeters: 'millimeter', degrees: 'degree', gigabytes: 'gigabyte', megabytes: 'megabyte', terabytes: 'terabyte',
  megawatts: 'megawatt', gigawatts: 'gigawatt', pounds: 'pound', feet: 'foot' };

const esc = (s) => s.replace(/[.*+?^${}()|[\]\\/]/g, '\\$&');
const CUR = Object.keys(CURRENCIES).map(esc).join('|');
const SCALE = 'trillion|billion|million|thousand|tn|bn|mn|mm|[kmbt]';
const MONTH = String.raw`\b(?:January|February|March|April|May|June|July|August|September|October|November|December`
  + String.raw`|Jan|Feb|Mar|Apr|Jun|Jul|Aug|Sept|Sep|Oct|Nov|Dec)\.?`;
const UNIT = Object.keys(UNITS).sort((a, b) => b.length - a.length).map(esc).join('|');
const EN_PATTERN = new RegExp([
  String.raw`(?<cur>${CUR})\s?(?<camt>${NUM})(?:\s?(?<cscale>${SCALE})\b)?`,
  String.raw`(?<ra>${NUM})\s?(?:-|–|to)\s?(?<rb>${NUM})\s?(?<rpct>%)`,
  String.raw`(?<pct>-?(?:${NUM}))\s?%`,
  String.raw`(?<month>${MONTH})\s(?<day>\d{1,2})(?!\d|,\d)(?:st|nd|rd|th)?`,
  String.raw`(?<h>\d{1,2}):(?<mi>\d{2})(?:\s?(?<ampm>[ap])\.?m\.?)?`,
  String.raw`(?<ord>\d+)(?:st|nd|rd|th)\b`,
  String.raw`(?<decade>1[1-9]\d0|20[0-9]0)s\b`,
  String.raw`(?<ya>1[1-9]\d{2}|20\d{2})\s?(?:-|–)\s?(?<yb>1[1-9]\d{2}|20\d{2})(?!\d)`,
  String.raw`(?<mult>${NUM})\s?[x×](?![a-z])`,
  String.raw`(?<![A-Za-z])(?<samt>${NUM})(?<sscale>bn|mn|tn|[kmbKMB])\b`,
  String.raw`(?<uamt>${NUM})\s?(?<unit>${UNIT})(?![A-Za-z])`,
  String.raw`#(?<hash>\d+)`,
  String.raw`(?<frac>\d+\/\d+)`,
  String.raw`(?<ra2>${NUM})\s?–\s?(?<rb2>${NUM})`,
  String.raw`(?<year>(?<![\d.,$])(?:1[1-9]\d{2}|20\d{2})(?![\d%]|\.\d|,\d))`,
  String.raw`(?<neg>(?<![\w.])-)?(?<num>${NUM})`,
].join('|'), 'g');

const isOne = (amount) => Number(amount.replace(/,/g, '')) === 1;

function enSpeak(m) {
  const g = m.groups;
  if (g.cur) {
    const [one, many] = CURRENCIES[g.cur];
    const amount = g.camt;
    if (g.cscale) return `${enNumber(amount)} ${SCALE_WORDS[g.cscale.toLowerCase()]} ${many}`;
    if (/^[\d,]+\.\d\d$/.test(amount) && !amount.endsWith('.00') && one === 'dollar') {
      const [whole, cents] = amount.replace(/,/g, '').split('.');
      return `${enNumber(whole)} ${Number(whole) === 1 ? one : many} and ${enNumber(cents)} cents`;
    }
    return `${enNumber(amount)} ${isOne(amount) ? one : many}`;
  }
  if (g.rpct) return `${enNumber(g.ra)} to ${enNumber(g.rb)} percent`;
  if (g.pct) return (g.pct.startsWith('-') ? 'minus ' : '') + `${enNumber(g.pct.replace(/^-/, ''))} percent`;
  if (g.month) return `${MONTHS[g.month.replace(/\.$/, '').toLowerCase().slice(0, 3)]} ${ordinal(Number(g.day))}`;
  if (g.h) {
    const hour = Number(g.h);
    const minute = Number(g.mi);
    if (hour > 24 || minute > 59) return null;
    const words = enNumber(String(hour)) + (minute === 0 && !g.ampm ? " o'clock" : minute === 0 ? ''
      : minute < 10 ? ` oh ${enNumber(String(minute))}` : ` ${enNumber(String(minute))}`);
    return words + (g.ampm ? ` ${g.ampm} m` : '');
  }
  if (g.ord) return ordinal(Number(g.ord));
  if (g.decade) {
    const words = enYear(g.decade);
    return words.endsWith('y') ? words.slice(0, -1) + 'ies' : words + 's';
  }
  if (g.ya) return `${enYear(g.ya)} to ${enYear(g.yb)}`;
  if (g.mult) return `${enNumber(g.mult)} times`;
  if (g.samt) return `${enNumber(g.samt)} ${SCALE_WORDS[g.sscale.toLowerCase()]}`;
  if (g.uamt) {
    let [first, ...rest] = UNITS[g.unit].split(' ');
    if (isOne(g.uamt)) first = SINGULAR[first] ?? first;
    return `${enNumber(g.uamt)} ${first}${rest.length ? ' ' + rest.join(' ') : ''}`;
  }
  if (g.hash) return `number ${enNumber(g.hash)}`;
  if (g.frac) {
    const [a, b] = g.frac.split('/');
    return FRACTIONS[g.frac] ?? `${enNumber(a)} ${enNumber(b)}`;
  }
  if (g.ra2) return `${enNumber(g.ra2)} to ${enNumber(g.rb2)}`;
  if (g.year) return enYear(g.year);
  return (g.neg ? 'minus ' : '') + enNumber(g.num);
}

export function normalize(display) {
  return rewrite(display, EN_PATTERN, enSpeak);
}
